### PR TITLE
rgw: clean-up -- use enum class for stats category

### PR DESCRIPTION
**Pull request by ivancich**
The stats entries for rgw buckets has a category, which used a combination of `uint8_t` and `enum RGWObjClass`. Clean this up by converting `RGWObjClass` to an `enum class` and using that throughout. This provides type safety and better code clarity.

### CODE REVIEW

**Comment by ivancich** [comment]
Waiting for https://github.com/ceph/ceph/pull/25333 to merge and for some testing before I'll remove DNM and add reviewers.

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -523,7 +523,8 @@ static int check_index(cls_method_context_t hctx, struct rgw_bucket_dir_header *
         CLS_LOG(1, "ERROR: rgw_bucket_list(): failed to decode entry, key=%s\n", kiter->first.c_str());
         return -EIO;
       }
-      struct rgw_bucket_category_stats& stats = calc_header->stats[entry.meta.category];
+      struct rgw_bucket_category_stats& stats =
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw.cc]: why the C-style struct notation here?

File: src/cls/rgw/cls_rgw_types.cc [hunk 2]
```diff
@@ -39,7 +41,7 @@ void cls_rgw_obj_key::decode_json(JSONObj *obj) {
 void rgw_bucket_dir_entry_meta::generate_test_instances(list<rgw_bucket_dir_entry_meta*>& o)
 {
   rgw_bucket_dir_entry_meta *m = new rgw_bucket_dir_entry_meta;
-  m->category = 1;
+  m->category = RGWObjCategory::Main;
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_types.cc]: cool;  although what does "Main" mean?

File: src/cls/rgw/cls_rgw_types.h [hunk 3]
```diff
@@ -91,8 +91,17 @@ struct rgw_bucket_pending_info {
 };
 WRITE_CLASS_ENCODER(rgw_bucket_pending_info)
 
+
+enum class RGWObjCategory : uint8_t {
+  None      = 0,
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_types.h]: could we annotate (//) with the meaning/intent of each category?

File: src/cls/rgw/cls_rgw_client.cc [hunk 4]
```diff
@@ -152,8 +152,9 @@ int CLSRGWIssueSetTagTimeout::issue_op(int shard_id, const string& oid)
   return issue_bucket_set_tag_timeout_op(io_ctx, oid, tag_timeout, &manager);
 }
 
-void cls_rgw_bucket_update_stats(librados::ObjectWriteOperation& o, bool absolute,
-                                 const map<uint8_t, rgw_bucket_category_stats>& stats)
+void cls_rgw_bucket_update_stats(librados::ObjectWriteOperation& o,
+				 bool absolute,
+                                 const map<RGWObjCategory, rgw_bucket_category_stats>& stats)
 {
   struct rgw_cls_bucket_update_stats_op call;
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_client.cc]: same here;  there must be a lot of this

File: src/cls/rgw/cls_rgw.cc (hunk 1 above)
Comment (ivancich) [reply src/cls/rgw/cls_rgw.cc]: Don't know! Cleaned it up across the check_index function. Looks like it's wide-spread in this file. Maybe another PR can clean all that up.

File: src/cls/rgw/cls_rgw_types.cc (hunk 2 above)
Comment (ivancich) [reply src/cls/rgw/cls_rgw_types.cc]: Added documentation in src/cls/rgw/cls_rgw_types.h.

File: src/cls/rgw/cls_rgw_types.h (hunk 3 above)
Comment (ivancich) [reply src/cls/rgw/cls_rgw_types.h]: Good idea; take a look and see if it's adequate. I did audit all uses of each of the 4 enums.

**Comment by ivancich** [comment]
Still waiting for https://github.com/ceph/ceph/pull/25333 to merge and then I'll remove DNM and add needs-qa (provided concerns adequately addressed).

**Comment by ivancich** [comment]
Building in preparation for a teuthology rgw suite run.
https://shaman.ceph.com/builds/ceph/wip-rgw-category-cleanup/

**Comment by ivancich** [comment]
Started an rgw-suite on teuthology:
http://pulpito.ceph.com/ivancich-2018-12-14_15:32:15-rgw-wip-rgw-category-cleanup-distro-basic-smithi/

File: src/cls/rgw/cls_rgw.cc [hunk 5]
```diff
@@ -2237,7 +2242,8 @@ static int rgw_bi_get_op(cls_method_context_t hctx, bufferlist *in, bufferlist *
       encode_olh_data_key(op.key, &idx);
       break;
     default:
-      CLS_LOG(10, "%s(): invalid key type encoding: %d", __func__, op.type);
+      CLS_LOG(10, "%s(): invalid key type encoding: %d",
+	      __func__, int(op.type));
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw.cc]: I skipped past this before--just to ask the naive question, would it make sense to make a human-readable print of the enumeration type?

File: src/cls/rgw/cls_rgw_types.h [hunk 6]
```diff
@@ -91,8 +91,26 @@ struct rgw_bucket_pending_info {
...
+// differentiate their associated statistics (bucket stats, and in
+// some cases user stats)
+enum class RGWObjCategory : uint8_t {
+  None      = 0,  // b-i entries for delete markers; also used in
+                  // testing and for default values in default
+                  // constructors
+
+  Main      = 1,  // b-i entries for standard objs
+
+  Shadow    = 2,  // presumfably intended for multipart shadow
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_types.h]: presumably (sp)?

File: src/cls/rgw/cls_rgw_client.cc [hunk 7]
```diff
@@ -122,7 +122,7 @@ static bool issue_bucket_index_clean_op(librados::IoCtx& io_ctx,
 static bool issue_bucket_set_tag_timeout_op(librados::IoCtx& io_ctx,
     const string& oid, uint64_t timeout, BucketIndexAioManager *manager) {
   bufferlist in;
-  struct rgw_cls_tag_timeout_op call;
+  rgw_cls_tag_timeout_op call;
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_client.cc]: yay

**Comment by mattbenjamin** [comment]
(adding needs-qa--can be cleared if you have an rgw suite run)

**Pull request merged**